mythril/support/signatures.py

```diff
@@ -1,44 +1,244 @@
+                    self.release(force=True)  # cleanup old lockfile > 5mins
+
+                time.sleep(0.5)  # busywait is evil
+                continue
+
+        raise Exception("SimpleFileLock: timeout hit. failed to aquire lock: %s"% (time.time()-self.lockfile.stat().st_mtime))
+
+    def release(self, force=False):
+        if not force and not self.locked:
+            raise Exception("SimpleFileLock: aquire lock first")
+
+        try:
+            self.lockfile.unlink()  # might throw if we force unlock and the file gets removed in the meantime. TOCTOU
+        except FileNotFoundError as fnfe:
+            logging.warning("SimpleFileLock: release(force=%s) on unavailable file. race? %r" % (force, fnfe))
+
+        self.locked = False
+
+
+
+class SignatureDb(object):
+
+    def __init__(self, enable_online_lookup=True):
+        """
+        Constr
+        :param enable_online_lookup: enable onlien signature hash lookup
+        """
+        self.signatures = {}  # signatures in-mem cache
+        self.signatures_file = None
+        self.signatures_file_lock = None
+        self.enable_online_lookup = enable_online_lookup  # enable online funcsig resolving
+        self.online_lookup_miss = set()  # temporarily track misses from onlinedb to avoid requesting the same non-existent sighash multiple times
+        self.online_directory_unavailable_until = 0  # flag the online directory as unavailable for some time
+
+    def open(self, path=None):
+        """
+        Open a function signature db from json file
+
+        :param path: specific path to signatures.json; default mythril location if not specified
+        :return: self
+        """
+        if not path:
+            #  try default locations
+            try:
+                mythril_dir = os.environ['MYTHRIL_DIR']
+            except KeyError:
+                mythril_dir = os.path.join(os.path.expanduser('~'), ".mythril")
+            path = os.path.join(mythril_dir, 'signatures.json')
+
+        self.signatures_file = path  # store early to allow error handling to access the place we tried to load the file
+
+        if not os.path.exists(path):
+            logging.debug("Signatures: file not found: %s" % path)
+            raise FileNotFoundError("Missing function signature file. Resolving of function names disabled.")
+
+        self.signatures_file_lock = SimpleFileLock(self.signatures_file)  # lock file to prevent concurrency issues
+        self.signatures_file_lock.aquire()  # try to aquire it within the next 10s
+
+        with open(path, 'r') as f:
+            sigs = json.load(f)
+
+        self.signatures_file_lock.release()  # release lock
+
+        # normalize it to {sighash:list(signatures,...)}
+        for sighash, funcsig in sigs.items():
+            if isinstance(funcsig, list):
+                self.signatures = sigs  # keep original todo: tintinweb - super hacky. make sure signatures.json is initially in correct format fixme
```

Comment:
Maybe remove/change the todo. As this might not make sense with a new persistence method

Reply:
this is more or less a reminder that we need to change the default `signatures.json` format to `{hash:[funcsig,...]}`. I'll remove the todo as - as you noted - this will have to be refactored anyway when switching to a new db backend.